geode-core/src/main/java/org/apache/geode/internal/net/SCAdvancedSocketCreator.java

```diff
@@ -83,8 +86,33 @@ public Socket connect(HostAndPort addr, int timeout,
       if (optionalWatcher != null) {
         optionalWatcher.beforeConnect(socket);
       }
-      socket.connect(sockaddr, Math.max(timeout, 0));
-      coreSocketCreator.configureClientSSLSocket(socket, timeout);
+
+      final InetSocketAddress sniProxyAddress =
+          coreSocketCreator.getSslConfig().getSniProxyAddress();
+      final boolean provideSNI = sniProxyAddress != null;
+
+      if (provideSNI) {
+        logger.error("DEBUG CREATING SSL PROXY SOCKET FOR {}", addr.getHostName());
+        socket.connect(sniProxyAddress, Math.max(timeout, 0));
+        coreSocketCreator.configureClientSSLSocket(socket, timeout, sslParameters -> {
+          final List<SNIServerName> sniHostNames = new ArrayList<>(1);
+          sniHostNames.add(new SNIHostName(addr.getHostName()));
+          sslParameters.setServerNames(sniHostNames);
+        });
+      } else {
+        final InetSocketAddress sockaddr;
+        InetSocketAddress sockaddrTemp = addr.getSocketInetAddress();
+        if (sockaddrTemp.getAddress() == null) {
+          final InetAddress address = InetAddress.getByName(sockaddrTemp.getHostName());
+          sockaddr = new InetSocketAddress(address, sockaddrTemp.getPort());
+        } else {
+          sockaddr = sockaddrTemp;
+        }
+        socket.connect(sockaddr, Math.max(timeout, 0));
+        coreSocketCreator.configureClientSSLSocket(socket, timeout, _ignored -> {
```

Comment:
Why not provide the SNI regardless of proxy?

Reply:
If I understand your question, I chose to pass a lambda to `configureClientSSLSocket()` rather than passing in a (possibly-`null` SNI hostname).

Reply:
My question is, There is always a host name so why not always set the SNI header?

Reply:
Oh I see what you mean. I'd rather not add that header, which is superfluous in the common case i.e. no proxy. But yeah like you, I think functionally that could work.

Reply:
I am thinking from the standpoint that everything is the same except for where it can't. So in this case the initialization of the SSL socket is identical regardless of the proxy. Less moving parts.